src/jdk.jfr/share/classes/jdk/jfr/internal/util/ParsedCPUThrottle.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2024, SAP SE. All rights reserved.
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.jfr.internal.util;
+
+import jdk.jfr.internal.settings.CPUThrottleSetting;
+
+/**
+ * A rate or fixed period, see {@link jdk.jfr.internal.Rate}
+ */
+public sealed interface ParsedCPUThrottle {
```

Comment:
This seems complicated. Why isn't possible to keep Rate class as is (no renaming) and then only create a TimespanRate class and do something like this:

`public record TimespanRate(double rate) implements ParsedCPUThrottle {

    public static TimespanRate of(String text) {
        Rate r = Rate.of(text);
        if (r != null) {
            double rate = r.amount() * ...
            return new TimespanRate(rate)
        } else {
            long v = ValueParser.parseTimespanWithInfinity(text);
            double rate = v ...
            return new TimespanRate(rate);
        }
    }
    
    public default boolean isHigher(ParsedCPUThrottle that) {
        return rate() > that.rate();
    }
`

Reply:
I can try to implement. I thought it would be easier to comprehend this way.

Reply:
Since you need custom logic in CPUThrottleSetting, it might be easier to just have two distinct types and skip the common interface?

I haven't looked at it closely, it's just my impression. Maybe there should be a ThrottlePeriod(double period, boolean autoAdapt) with a factory method where it first checks if it can be created from Rate.of(text).

Reply:
You were right, the code is now far simpler. I dropped the renaming of Rate and created a TimespanRate(double rate, boolean autoadapt).